stackrox/stackrox
# Delete obsolete schema and functions for images service in UI

## Description

Follow up long after /main/vulnerability-management/images superseded /main/images and therefore GraphQL query superseded REST request

* https://github.com/stackrox/stackrox/blob/master/proto/api/v1/image_service.proto
* https://github.com/stackrox/stackrox/blob/master/proto/storage/image.proto

### Analysis

* `fetchImages` has 1 call in Containers/Policies/Wizard/Step4/PolicyScopeForm.tsx
* `fetchImageCount` has 0 calls
* `getWatchedImages` has 1 call in Containers/VulnMgmt/List/Images/WatchedImagesDialog.tsx
* `unwatchImage` has 1 call in Containers/VulnMgmt/List/Images/WatchedImagesDialog.tsx
* `watchImage` has 1 call in Containers/VulnMgmt/List/Images/WatchedImagesDialog.tsx

### Changed files

1. Edit Containers/Policies/Wizard/Step4/PolicyScopeForm.tsx
    * Rename `Image` as `ListImage`

2. Edit Containers/VulnMgmt/List/Images/WatchedImagesDialog.tsx
    * Replace `{ name: string }[]` with `WatchedImage[]` in type of `useState` hook
    * Replace `image` and `image?.normalizedName` with `normalizedName`

3. Rewrite services/ImagesService.js as TypeScript
    * Delete unneeded arguments and schema from `fetchImages` function and rename as `getImages` in this case intentionally not parallel to `ListImages` in backend service which was an early naming convention when plural request returned subset of information in singular request
    * Delete `fetchImageCount` function
    * Replace `axios(options)` calls with `axios.method<Parameter>` calls
    * Replace `{ normalizedName }` with `normalizedName` as return value of `watchImage` function (and make corresponding changes in WatchedImagesDialog.tsx file)

4. Edit services/schemas.js
    * Delete `image` schema

5. Edit types/image.proto.ts
    * Replace `Image` with `ListImage` type
    * Add `WatchedImage` type

### Residue

1. Rewrite 3 functions in ClustersService to remove need for `cluster` schema.
2. Remove `normalizr` from dependencies.

## Checklist

## Testing Performed

1. `yarn build` in ui
2. `wc build/static/js/*.chunk.js` in ui
    * branch - master: -1515 = 10398896 - 10400411
3. `yarn start` in ui
4. Test `getImages` function:
    * Visit /main/policy-management/policies
    * Click **Create policy** button
    * Continue to step 4: see GET /v1/images request
        ![policies-images-network](https://user-images.githubusercontent.com/11862657/179030558-a9f1a018-4ccd-44bd-a337-ab03f9294814.png)

        compare response to temporary `console.log` in PolicyScopeForm.tsx file
        ![policies-iimages-console](https://user-images.githubusercontent.com/11862657/179030598-9a90e474-469f-4da8-b9a9-08a16e60b15b.png)

5. Test watched image functions:
    * Visit /main/vulnerability-management/images
    * Click **Manage Scanning of Watched Images** button whose icon looks like bulleted list: see GET /v1/watchedimages request and `{watchedImages: []}` response
    * In **Manage Watched Images** modal, type **quay.io/stackrox-io/scanner-db:2.25.13** which has extra incorrect 3 at the end, and then click **Add Image**
        see POST /v1/watchedimages request
        with payload `{name: "quay.io/stackrox-io/scanner-db:2.25.13"}`
        with response
        ```js
        {
            errorMessage: "failed to scan image: image enrichment error: error getting metadata for image: quay.io/stackrox-io/scanner-db:2.25.13 error: getting metadata from registry: \"Public Quay.io\": Failed to get the manifest digest : Head \"https://quay.io/v2/stackrox-io/scanner-db/manifests/2.25.13\": http: non-successful response (status=404 body=\"\")",
            errorType: "SCAN_FAILED",
            normalizedName: ""
        }
        ```
        ![watched-errorType_SCAN_FAILED](https://user-images.githubusercontent.com/11862657/179030696-b1c12650-a46c-4c20-b878-00f100e3b06c.png)

    * In **Manage Watched Images** modal, type **quay.io/stackrox-io/scanner-db:2.25.1** which is correct, and then click **Add Image**
        see POST /v1/watchedimages request
        with payload `{name: "quay.io/stackrox-io/scanner-db:2.25.1"}`
        with response
        ```js
        {
            errorMessage: "",
            errorType: "NO_ERROR",
            normalizedName: "quay.io/stackrox-io/scanner-db:2.25.1"
        }
        ```
        ![watched-errorType_NO_ERROR](https://user-images.githubusercontent.com/11862657/179030748-991481e9-14ec-42f4-8e4e-a504ab5b2f46.png)

    * Click **Return to Image List**: see **Scanning via watch tag** link in **Image status** column (not sure why there is a pair of images)
        see GET /v1/watchedimages request with `{watchedImages: [{name: "quay.io/stackrox-io/scanner:2.25.1"}]}` response
        ![watched-images_Scanning_via_watch_tag](https://user-images.githubusercontent.com/11862657/179030929-e64fe1dd-d4c0-4c73-962a-7a3ee94ad6b7.png)


    * Click **Scanning via watch tag** link, click **Remove Watch** for **quay.io/stackrox-io/scanner-db:2.25.1**
        see DELETE /v1/watchedimages?name=quay.io/stackrox-io/scanner-db:2.25.1 request with `{}` response
        see GET /v1/watchedimages request with `{watchedImages: []}` response
    * Click **Return to Image List**: see no link in **Image status** column
